src/sentry/snuba/discover.py

```diff
@@ -0,0 +1,230 @@
+
+
+def transform_results(result, translated_columns):
+    # Translate back columns that were converted to snuba format
+    for col in result["meta"]:
+        col["name"] = translated_columns.get(col["name"], col["name"])
+
+    def get_row(row):
+        return {translated_columns.get(key, key): value for key, value in row.items()}
+
+    if len(translated_columns):
+        result["data"] = [get_row(row) for row in result["data"]]
+
+    # TODO(mark) add zerofill for timeseries data.
+    return result
+
+
+def query(selected_columns, query, params, orderby=None, referrer=None):
+    """
+    High-level API for doing arbitrary user queries against events.
+
+    This function operates on the public event schema and
+    virtual fields/aggregate functions for selected columns and
+    conditions are supported through this function.
+
+    The resulting list will have all internal field names mapped
+    back into their public schema names.
+
+    selected_columns (Sequence[str]) List of public aliases to fetch.
+    query (str) Filter query string to create conditions from.
+    params (Dict[str, str]) Filtering parameters with start, end, project_id, environment
+    orderby (str|Sequence[str]) The field to order results by.
+    """
+    snuba_filter = get_filter(query, params)
+
+    # TODO(mark) Refactor the need for this translation shim once all of
+    # discover is using this module
+    snuba_args = {
+        "start": snuba_filter.start,
+        "end": snuba_filter.end,
+        "conditions": snuba_filter.conditions,
+        "filter_keys": snuba_filter.filter_keys,
+        "orderby": orderby,
+    }
+    snuba_args.update(resolve_field_list(selected_columns, snuba_args))
+
+    # Resolve the public aliases into the discover dataset names.
+    snuba_args = resolve_discover_aliases(snuba_args)
+    translated_columns = snuba_args.pop("translated_columns")
+
+    result = raw_query(
+        start=snuba_args.get("start"),
+        end=snuba_args.get("end"),
+        groupby=snuba_args.get("groupby"),
+        conditions=snuba_args.get("conditions"),
+        aggregations=snuba_args.get("aggregations"),
+        selected_columns=snuba_args.get("selected_columns"),
+        filter_keys=snuba_args.get("filter_keys"),
+        orderby=snuba_args.get("orderby"),
+        dataset=Dataset.Discover,
+        referrer=referrer,
+    )
+
+    return transform_results(result, translated_columns)
+
+
+def timeseries_query(selected_columns, query, params, rollup):
+    """
+    High-level API for doing arbitrary user timeseries queries against events.
+
+    This function operates on the public event schema and
+    virtual fields/aggregate functions for selected columns and
+    conditions are supported through this function.
+
+    This function is intended to only get timeseries based
+    results and thus requires the `rollup` parameter.
+    """
+    raise NotImplementedError
+
+
+def get_pagination_ids(event, query, params):
```

Comment:
Isn't this akin `get_next_event_id` and friends that are provided by eventstore?

Reply:
Yes, we had talked about having a discover layer method that does query string parsing and date range limiting. Currently that logic is in the endpoints and I'd like to get it out of there.

Reply:
How does this fan out into the 4 methods for getting first/last/next/prev? Wouldn't it be more a matter just calling your new `resolve_discover_aliases` methods to get the internal form and passing that on to the relevant eventstore function?

Reply:
I was thinking that this method would:

1. Resolve field aliases and generate 'reference event' conditions based on the provided event.
2. Do queries for the next/previous/latest/oldest events.
3. Return a dictionary of the eventids or none.

Reply:
Ah, so it's responsible for calling all 4 methods. Sounds good.